pkg/http/auth_middleware.go

```diff
@@ -0,0 +1,352 @@
+			authorizedError(w, err)
+			return
+		}
+
+		handler.ServeHTTP(w, rq)
+	})
+}
+
+func authorizedError(w http.ResponseWriter, err error) {
+	w.WriteHeader(http.StatusUnauthorized)
+	w.Header().Set("Content-Type", "application/xml")
+
+	xml := fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
+<Error>
+  <Code>Unauthorized</Code>
+  <Message>%s</Message>
+</Error>`, err.Error())
+
+	if _, err := w.Write([]byte(xml)); err != nil {
+		return
+	}
+}
+
+// sum256 calculate sha256 sum for an input byte array.
+func sum256(data []byte) []byte {
+	hash := sha256.New()
+	hash.Write(data)
+
+	return hash.Sum(nil)
+}
+
+// sumHMAC calculate hmac between two input byte array.
+func sumHMAC(key []byte, data []byte) []byte {
+	hash := hmac.New(sha256.New, key)
+	hash.Write(data)
+
+	return hash.Sum(nil)
+}
+
+func getHashedPayload(req http.Request) string {
+	hashedPayload := req.Header.Get("X-Amz-Content-Sha256")
+	if hashedPayload == "" {
+		// Presign does not have a payload, use S3 recommended value.
+		hashedPayload = unsignedPayload
+	}
+
+	return hashedPayload
+}
+
+func headerExists(key string, headers []string) bool {
+	for _, k := range headers {
+		if k == key {
+			return true
+		}
+	}
+
+	return false
+}
+
+// getHostAddr returns host header if available, otherwise returns host from URL
+func getHostAddr(req *http.Request) string {
+	host := req.Header.Get("host")
+	if host != "" && req.Host != host {
+		return host
+	}
+
+	if req.Host != "" {
+		return req.Host
+	}
+
+	return req.URL.Host
+}
+
+// Trim leading and trailing spaces and replace sequential spaces with one space, following Trimall()
+// in http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+func signV4TrimAll(input string) string {
+	// Compress adjacent spaces (a space is determined by
+	// unicode.IsSpace() internally here) to one space and return
+	return strings.Join(strings.Fields(input), " ")
+}
+
+func getCanonicalHeaders(req http.Request, signedHeaders string) string {
+	headers := []string{}
+	vals := make(map[string][]string)
+
+	for _, k := range strings.Split(signedHeaders, ";") {
+		vv := req.Header.Values(k)
+		headers = append(headers, strings.ToLower(k))
+		vals[strings.ToLower(k)] = vv
+	}
+
+	if !headerExists("host", headers) {
+		headers = append(headers, "host")
+	}
+
+	sort.Strings(headers)
+
+	var buf bytes.Buffer
+	// Save all the headers in canonical form <header>:<value> newline
+	// separated for each header.
+	for _, k := range headers {
+		buf.WriteString(k)
+		buf.WriteByte(':')
+
+		switch {
+		case k == "host":
+			buf.WriteString(getHostAddr(&req))
+			buf.WriteByte('\n')
+		default:
+			for idx, v := range vals[k] {
+				if idx > 0 {
+					buf.WriteByte(',')
+				}
+
+				buf.WriteString(signV4TrimAll(v))
+			}
+
+			buf.WriteByte('\n')
+		}
+	}
+
+	return buf.String()
+}
+
+func encodePath(pathName string) string {
+	var reservedObjectNames = regexp.MustCompile("^[a-zA-Z0-9-_.~/]+$")
```

Comment:
Could be initialised at `init()` or as a global `var`, as it only needs to happen once.